### PR TITLE
Attestation c

### PR DESCRIPTION
**Pull request by dorin-iancu**

### CODE REVIEW

File: c-version/storage.c [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+i64 _loadAttestorListLen()
+{
+    return int64storageLoad(TOTAL_ATTESTATORS_KEY, TOTAL_ATTESTATORS_KEY_LEN);
+}
+
+void _storeAttestorListLen(i64 len)
+{
+    int64storageStore(TOTAL_ATTESTATORS_KEY, TOTAL_ATTESTATORS_KEY_LEN, len);
```
Comment (andrei-marinica) [inline c-version/storage.c]: int64storageStore is deprecated
use smallIntStorageStoreUnsigned instead

File: c-version/storage.c [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+{
+    byte serialized[sizeof(u64)] = {};
+
+    _serializeu64(nonce, serialized);
+    storageStore(MAX_NONCE_DIFF_KEY, MAX_NONCE_DIFF_KEY_LEN, serialized, sizeof(u64));
+}
+
+i64 _loadAttestorListLen()
+{
+    return int64storageLoad(TOTAL_ATTESTATORS_KEY, TOTAL_ATTESTATORS_KEY_LEN);
```
Comment (andrei-marinica) [inline c-version/storage.c]: int64storageLoad is deprecated
use smallIntStorageLoadUnsigned instead

File: c-version/elrond/context.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+// Blockchain-related functions
+long long getBlockTimestamp();
+long long getBlockNonce();
+int getBlockHash(long long nonce, byte *hash);
+
+// Argument-related functions
+int getNumArguments();
+int getArgument(int argumentIndex, byte *argument);
+long long int64getArgument(int argumentIndex);
```
Comment (andrei-marinica) [inline c-version/elrond/context.h]: int64getArgument is deprecated
use smallIntGetUnsignedArgument instead (or smallIntGetSignedArgument, but I don't think you need signed numbers anywhere in the contract)

File: c-version/elrond/context.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// Account-related functions
+void getExternalBalance(const byte *address, byte *balance);
+int transferValue(const byte *destination, const byte *value, const byte *data, int length);
+
+// Storage-related functions
+int storageLoadLength(const byte *key, int keyLength);
+int storageStore(const byte *key, int keyLength, const byte *data, int dataLength);
+int storageLoad(const byte *key, int keyLength, byte *data);
+int int64storageStore(const byte *key, int keyLength, long long value);
+long long int64storageLoad(const byte *key, int keyLength);
```
Comment (andrei-marinica) [inline c-version/elrond/context.h]: int64storageStore and int64storageLoad are deprecated, we should remove them from here from all contracts

I'm posting here the small int EI from arwen:

// extern long long smallIntGetUnsignedArgument(void *context, int32_t id);
// extern long long smallIntGetSignedArgument(void *context, int32_t id);
//
// extern void smallIntFinishUnsigned(void* context, long long value);
// extern void smallIntFinishSigned(void* context, long long value);
//
// extern int32_t smallIntStorageStoreUnsigned(void *context, int32_t keyOffset, int32_t keyLength, long long value);
// extern int32_t smallIntStorageStoreSigned(void *context, int32_t keyOffset, int32_t keyLength, long long value);
// extern long long smallIntStorageLoadUnsigned(void *context, int32_t keyOffset, int32_t keyLength);
// extern long long smallIntStorageLoadSigned(void *context, int32_t keyOffset, int32_t keyLength);
//
// ... and 4 deprecated functions that we should no longer have in the headers

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 5]
```diff
@@ -10,3 +9,7 @@ Cargo.lock
 
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+# C buildchain artifacts
+*.ll
+*.o
```
Comment (sasurobert) [inline .gitignore]: empty line to end of file

File: c-version/helpers.c [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    for (int i = 0; i < n; i++)
+    {
+        cdest[i] = csrc[i];
+    }
+
+    return dest;
+}
+
+// fake memset
```
Comment (sasurobert) [inline c-version/helpers.c]: these are not fake. these are the actual implementation :) delete comment.

File: c-version/helpers.c [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    ADDRESS caller = {};
+    ADDRESS owner = {};
+
+    getCaller(caller);
+    getOwnerAddress(owner);
+
+    return _equal(caller, owner, sizeof(ADDRESS));
+}
+
+// fake memcpy
```
Comment (sasurobert) [inline c-version/helpers.c]: these are not fake. these are the actual implementation :) delete comment.

**Review by sasurobert — COMMENTED** [review]
Looks impressive!

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**